test/test_quantization.py

```diff
@@ -535,200 +535,204 @@ def checkQuantized(model):
+            cell_fp16 = model_fp16.lstm
+
+            assert type(cell_int8) == torch.nn.quantized.dynamic.LSTM, \
+                'torch.nn.LSTM should be converted to torch.nn.quantized.dynamic.LSTM after quantize_dynamic'
+            assert type(cell_fp16) == torch.nn.quantized.dynamic.LSTM, \
+                'torch.nn.LSTM should be converted to torch.nn.quantized.dynamic.LSTM after quantize_dynamic'
+
+            niter = 10
+            x = torch.tensor([[100, -155],
+                              [-155, 100],
+                              [100, -155]], dtype=torch.float).unsqueeze(0).repeat(niter, 1, 1)
+
+            h0_vals = [[-155, 100],
+                       [-155, 155],
+                       [100, -155]]
+
+            hx = torch.tensor(h0_vals, dtype=torch.float).unsqueeze(0)
+            cx = torch.tensor(h0_vals, dtype=torch.float).unsqueeze(0)
+
+            if isinstance(ref, torch.nn.LSTM):
+                hiddens = (hx, cx)
+
+            ref_out, ref_hid = ref(x, hiddens)
+
+            # Compare int8 quantized to unquantized
+            output_int8, final_hiddens_int8 = cell_int8(x, hiddens)
+
+            torch.testing.assert_allclose(output_int8, ref_out)
+            self.assertEqual(output_int8, ref_out)
+            for out_val, ref_val in zip(final_hiddens_int8, ref_hid):
+                torch.testing.assert_allclose(out_val, ref_val)
+
+            class ScriptWrapper(torch.nn.Module):
+                def __init__(self, cell):
+                    super(ScriptWrapper, self).__init__()
+                    self.cell = cell
+
+                def forward(self, x, hiddens):
+                    # type: (torch.Tensor, Tuple[torch.Tensor, torch.Tensor])
+                    # -> Tuple[torch.Tensor, Tuple[torch.Tensor, torch.Tensor]]
+                    return self.cell(x, hiddens)
+
+            # TODO: TorchScript overloads don't work without this wrapper
+            cell_script = torch.jit.script(ScriptWrapper(cell_int8))
+            out_script, hid_script = cell_script(x, hiddens)
+            self.assertEqual(len(out_script), len(ref_out))
+            for out_val, ref_val in zip(out_script, ref_out):
+                torch.testing.assert_allclose(out_val, ref_val)
+
+            # Test save/load
+            b = io.BytesIO()
+            torch.jit.save(cell_script, b)
+            b.seek(0)
+            loaded = torch.jit.load(b)
+            out_loaded, hid_loaded = loaded(x, hiddens)
+            for loaded_val, ref_val in zip(out_loaded, ref_out):
+                torch.testing.assert_allclose(loaded_val, ref_val)
 
-        torch.testing.assert_allclose(output_fp16, ref_out)
-        self.assertEqual(output_fp16, ref_out)
-        for out, ref_val in zip(final_hiddens_fp16, ref_hid):
-            torch.testing.assert_allclose(out, ref_val)
+            # Compare fp16 quantized to unquantized
+            output_fp16, final_hiddens_fp16 = cell_fp16(x, hiddens)
```

Comment:
When `qengine` is `qnnpack`, can we pass the test for FP16 dynamic quantized LSTM? I don't think QNNPACK currently support FP16.

Reply:
In that case if USE_FBGEMM is set to true then it will just execute fbgemm test. I don't think we call `gather_quantized_params_dynamic` for fp16 test, do we?

Reply:
For FP16, we will call `gather_quantized_params_fp16`:
https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/RNN.cpp#L1226-L1231

Reply:
@jianyuh then it should not depend on what qengine is set and follow the older code path.